modules/spechla/1.0/spechla.smk

```diff
@@ -0,0 +1,310 @@
+    params:
+        outdir = CFG["dirs"]["spechla"] + "{seq_type}--{genome_build}/{sample_id}",
+    shell:
+        op.as_one_line("""
+            touch {output.freq_list} && 
+            ls {params.outdir}/*_freq.txt > {output.freq_list} || true
+        """)
+
+# Fetch purity and ploidy from a table
+# Table should be formatted as "sample\tploidy\tpurity"
+def _get_sample_ploidy(wildcards):
+    CFG = config["lcr-modules"]["spechla"]
+    purity_ploidy_file = CFG["options"]["purity_ploidy_file"] 
+    with open(purity_ploidy_file, 'r') as f:
+        for line in f.readlines():
+            line = line.rstrip('\n').rstrip('\r')
+            if str({wildcards.sample_id}) in line:
+                cols = line.split('\t')
+                ploidy = cols[2]
+                valid_ploidy = ploidy.isnumeric()
+                if valid_ploidy:
+                    return ploidy
+                else:
+                    return 2
+            else:
+                return 2
+
+def _get_sample_purity(wildcards):
+    CFG = config["lcr-modules"]["spechla"]
+    purity_ploidy_file = CFG["options"]["purity_ploidy_file"] 
+    with open(purity_ploidy_file, 'r') as f:
+        for line in f.readlines():
+            line = line.rstrip('\n').rstrip('\r')
+            if str({wildcards.sample_id}) in line:
+                cols = line.split('\t')
+                purity = cols[3]
+                valid_purity = purity.isnumeric()
+                if valid_purity:
+                    return purity
+                else:
+                    return 1
+            else:
+                return 1
+
+rule _spechla_loh:
+    input:
+        hla_results = str(rules._spechla_hla_typing.output.hla_results),
+        cleanup = str(rules._spechla_cleanup.output.done),
+        freq_list = str(rules._spechla_generate_freqlist.output.freq_list)
+    output:
+        loh_info = CFG["dirs"]["loh"] + "{seq_type}--{genome_build}/{sample_id}/merge.hla.copy.txt"
+    log:
+        stderr = CFG["logs"]["loh"] + "{seq_type}--{genome_build}/{sample_id}/loh.stderr.log"
+    params:
+        scriptdir = CFG["dirs"]["inputs"] + "SpecHLA",
+        sample_id = "{sample_id}",
+        purity = _get_sample_purity,
+        ploidy = _get_sample_ploidy,
+        outdir = CFG["dirs"]["loh"] + "{seq_type}--{genome_build}/{sample_id}/"
+    conda: CFG["conda_envs"]["spechla"]
+    threads:
+        CFG["threads"]["spechla"]
+    resources:
+        **CFG["resources"]["spechla"] 
+    shell:
+        op.as_one_line("""
+        perl {params.scriptdir}/script/cal.hla.copy.pl 
+        -purity {params.purity}
+        -ploidy {params.ploidy}
+        -S {params.sample_id} 
+        -F {input.freq_list}
+        -T {input.hla_results}
+        -O {params.outdir} > {log.stderr} 2>&1
+        """)
+
+# Symlinks the final output files into the module results directory (under '99-outputs/')
+rule _spechla_output_txt:
```

Comment:
Similar to how you've grouped the input symlinks, you could group the output symlinks with the previous rule.